src/main/scala/org/apache/spark/sql/execution/datasources/greenplum/GreenplumUtils.scala

```diff
@@ -60,71 +61,172 @@ object GreenplumUtils extends Logging {
+      } else {
+        values(i) = "NULL"
+      }
+      i += 1
+    }
+    (values.mkString(delimiter) + "\n").getBytes("UTF-8")
+  }
+
+  def convertValue(str: String, delimiter: Char): String = {
+    str.flatMap {
+      case '\\' => "\\\\"
+      case '\n' => "\\n"
+      case '\r' => "\\r"
+      case `delimiter` => s"\\$delimiter"
+      case c => s"$c"
+    }
+  }
+
   /**
    * https://www.postgresql.org/docs/9.2/sql-copy.html
    *
+   * Copy data to greenplum in a single transaction.
+   *
    * @param df the [[DataFrame]] will be copy to the Greenplum
    * @param schema the table schema in Greemnplum
    * @param options Options for the Greenplum data source
    */
-  def copyToGreenplum(df: DataFrame, schema: StructType, options: GreenplumOptions): Unit = {
-    val valueConverters: Array[(Row, Int) => String] =
-      schema.map(s => makeConverter(s.dataType, options)).toArray
+  def transactionalCopy(
+      df: DataFrame,
+      schema: StructType,
+      options: GreenplumOptions): Unit = {
+    val randomString = UUID.randomUUID().toString.filterNot(_ == '-')
+    val suffix = "sparkGpTmp"
+    val tempTable = s"${options.table}_${randomString}_$suffix"
+    val strSchema = JdbcUtils.schemaString(df, options.url, options.createTableColumnTypes)
+    val createTempTbl = s"CREATE TABLE $tempTable ($strSchema) ${options.createTableOptions}"
+
+    val conn = JdbcUtils.createConnectionFactory(options)()
+
+    try {
+      executeStatement(conn, createTempTbl)
+      val accumulator = df.sparkSession.sparkContext.longAccumulator("copySuccess")
+      val partNum = df.rdd.getNumPartitions
 
-    def convertRow(row: Row): Array[Byte] = {
-      var i = 0
-      val values = new Array[String](schema.length)
-      while (i < schema.length) {
-        if (!row.isNullAt(i)) {
-          values(i) = convertValue(valueConverters(i).apply(row, i))
-        } else {
-          values(i) = "NULL"
-        }
-        i += 1
+      df.foreachPartition { rows =>
+        copyPartition(rows, options, schema, tempTable, Some(accumulator))
       }
-      (values.mkString(options.delimiter) + "\n").getBytes("UTF-8")
-    }
 
-    def convertValue(str: String): String = {
-      assert(options.delimiter.length == 1, "The delimiter should be a single character.")
-      val delimiter = options.delimiter.charAt(0)
-      str.flatMap {
-        case '\\' => "\\\\"
-        case '\n' => "\\n"
-        case '\r' => "\\r"
-        case `delimiter` => s"\\$delimiter"
-        case c => s"$c"
+      if (accumulator.value == partNum) {
+        val dropTbl = s"DROP TABLE IF EXISTS ${options.table}"
+        executeStatement(conn, dropTbl)
+
+        val renameTempTbl = s"ALTER TABLE $tempTable RENAME TO ${options.table}"
+        executeStatement(conn, renameTempTbl)
+      } else {
+        val dropTempTbl = s"DROP TABLE $tempTable"
+        executeStatement(conn, dropTempTbl)
```

Comment:
we should handle exception here, so that the next exception will be throw properly

Reply:
Thanks for your valuable advice.